erdos/src/dataflow/state.rs

```diff
@@ -1,38 +1,81 @@
 //! Structures and traits for states added to streams.
 
 use crate::dataflow::Timestamp;
-
-// TODO: keep around messages. Add an iterator over messages.
-// Add set_timestamp and set_access_context to State.
-
-/// Trait that must be implemented by stream state.
-pub trait State: 'static + Clone + Send + Sync {}
-impl<T: 'static + Clone + Send + Sync> State for T {}
+use std::collections::HashMap;
 
 // TODO (Sukrit): Do these state traits also require a way to read the state for a given timestamp?
 
 /// Trait that must be implemented by a state structure that is used in a Parallel operator.
 /// This structure must implement a `commit` method that commits the final state for a given
 /// timestamp `t`.
-pub trait StateT: 'static + Send + Sync {
+pub trait State: 'static + Send + Sync {
+    type Item: Default;
+
     fn commit(&mut self, timestamp: &Timestamp);
 
-    fn get_last_committed_timestamp(&self) -> Timestamp;
+    fn last_committed_timestamp(&self) -> Timestamp;
+
+    fn at(&mut self, timestamp: &Timestamp) -> Option<&mut Self::Item>;
 }
 
-impl StateT for () {
+impl State for () {
+    type Item = ();
+
     fn commit(&mut self, _timestamp: &Timestamp) {}
 
-    fn get_last_committed_timestamp(&self) -> Timestamp {
+    fn last_committed_timestamp(&self) -> Timestamp {
         Timestamp::Bottom
     }
+
+    fn at(&mut self, _timestamp: &Timestamp) -> Option<&mut Self::Item> {
+        None
+    }
+}
+
+pub struct TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    state: HashMap<Timestamp, S>,
+    last_committed_timestamp: Timestamp,
+}
+
+impl<S> TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    pub fn new() -> Self {
+        Self {
+            state: HashMap::new(),
+            last_committed_timestamp: Timestamp::Bottom,
+        }
+    }
+}
+
+impl<S> State for TimeVersionedState<S>
+where
+    S: 'static + Default + Send + Sync,
+{
+    type Item = S;
+
+    fn commit(&mut self, timestamp: &Timestamp) {
+        self.last_committed_timestamp = timestamp.clone();
+    }
+
+    fn last_committed_timestamp(&self) -> Timestamp {
+        self.last_committed_timestamp.clone()
+    }
+
+    fn at(&mut self, timestamp: &Timestamp) -> Option<&mut Self::Item> {
+        Some(self.state.entry(timestamp.clone()).or_default())
+    }
 }
 
 /// Trait that must be implemented by a state structure that is used in a Sequential operator.
 /// This state structure must implement an `append` method that enables message callbacks to add
 /// intermediate state to the structure, and a `commit` method that commits the final state for a
 /// given timestamp t.
```

Comment:
Could you move parts of this documentation to the methods?

Also, can you change `get_last_committed_timestamp` to `last_committed_timestamp`?